### PR TITLE
Schedule a rendering update directly instead of going via GraphicsLayerUpdater

### PR DESCRIPTION
**Pull request by mattwoodrow**
#### f06fb535234af9c8789c3b01ae0ee10d8d22995a
<pre>
Schedule a rendering update directly instead of going via GraphicsLayerUpdater
<a href="https://bugs.webkit.org/show_bug.cgi?id=257247">https://bugs.webkit.org/show_bug.cgi?id=257247</a>
&lt;rdar://problem/109751273&gt;

Reviewed by Simon Fraser.

GraphicsLayerUpdater just waits for displayDidRefresh, and then calls back into RenderLayerCompositor to schedule a rendering update.

I think we can just schedule the rendering update directly, for the same outcome.

* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::windowScreenDidChange):
* Source/WebCore/platform/graphics/GraphicsLayerClient.h:
* Source/WebCore/platform/graphics/GraphicsLayerUpdater.cpp: Removed.
* Source/WebCore/platform/graphics/GraphicsLayerUpdater.h: Removed.
* Source/WebCore/rendering/RenderLayerCompositor.cpp:
(WebCore::RenderLayerCompositor::RenderLayerCompositor):
(WebCore::RenderLayerCompositor::~RenderLayerCompositor):
(WebCore::RenderLayerCompositor::notifyFlushBeforeDisplayRefresh):
(WebCore::RenderLayerCompositor::destroyRootLayer):
(WebCore::RenderLayerCompositor::flushLayersSoon): Deleted.
(WebCore::RenderLayerCompositor::displayRefreshMonitorFactory): Deleted.
(WebCore::RenderLayerCompositor::windowScreenDidChange): Deleted.
* Source/WebCore/rendering/RenderLayerCompositor.h:

Canonical link: <a href="https://commits.webkit.org/264488@main">https://commits.webkit.org/264488@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/734bb440b42ba651481048e01a75e44a8253a96d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/7809 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/8084 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/8266 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/9454 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/7951 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/7815 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/10075 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/8001 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/10811 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/7947 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/9054 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/7152 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/9566 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/6363 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/7111 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/14760 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/7500 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/7238 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/10633 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/7731 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/6296 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/7053 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1859 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/11263 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/7464 "Built successfully") | | |

### CODE REVIEW

**Comment by smfr** [comment]
`GraphicsLayerUpdater` exists to trigger rendering updates _when nothing else does so_, when accelerated animations which affect tiled layers are running. So I think it needs to continue to exist as a `DisplayRefreshMonitorClient` in order to continually trigger updates.

To test this, have a large (and therefore tiled) compositing layer which is moved around by a CSS transform animation; even when nothing else is happening, we should be recomputing the tile coverage on every frame.

**Comment by mattwoodrow** [comment]
> `GraphicsLayerUpdater` exists to trigger rendering updates _when nothing else does so_, when accelerated animations which affect tiled layers are running. So I think it needs to continue to exist as a `DisplayRefreshMonitorClient` in order to continually trigger updates.
> 
> To test this, have a large (and therefore tiled) compositing layer which is moved around by a CSS transform animation; even when nothing else is happening, we should be recomputing the tile coverage on every frame.

I think we just need a delay to finish the current rendering update (if we're in one), and then schedule a rendering update (with layer flush) on the next `displayDidRefresh`.

I believe that's what we're doing now, just using `GraphicsLayerUpdater` as the mechanism to delay (and exit the current call stack), rather than a 0ms timer.

File: Source/WebCore/rendering/RenderLayerCompositor.cpp [hunk 1]
```diff
@@ -438,6 +439,8 @@ RenderLayerCompositor::~RenderLayerCompositor()
 #endif
 
     ASSERT(m_rootLayerAttachment == RootLayerUnattached);
+
+    m_updateRenderingTimer.stop();
```
Comment (smfr) [inline Source/WebCore/rendering/RenderLayerCompositor.cpp]: Shouldn't be necessary; timer destructors call stop().

File: Source/WebCore/rendering/RenderLayerCompositor.h [hunk 2]
```diff
@@ -179,6 +177,7 @@ class RenderLayerCompositor final : public GraphicsLayerClient, public GraphicsL
     // GraphicsLayers buffer state, which gets pushed to the underlying platform layers
     // at specific times.
     void notifyFlushRequired(const GraphicsLayer*) override;
+    void notifyFlushBeforeDisplayRefresh(const GraphicsLayer*) override;
```
Comment (smfr) [inline Source/WebCore/rendering/RenderLayerCompositor.h]: I'd like to rename this; the "before display refresh" isn't really accurate. It's more like "notify needs another rendering update". Maybe `notifySubsequentFlushRequired` or something.

**Review by smfr — APPROVED** [review]

**Pull request merged**